File: Python/Problem_078.py
def euler_78(mod: int = 1_000_000) -> int:
    p = [1]  # p[0] = 1
    n = 1

    while True:
        total = 0
        k = 1
        while True:
            g1 = k * (3 * k - 1) // 2
            if g1 > n:
                break
            g2 = k * (3 * k + 1) // 2

            sign = 1 if (k & 1) else -1

            total += sign * p[n - g1]
            if g2 <= n:
                total += sign * p[n - g2]

            k += 1

        total %= mod
        p.append(total)

        if total == 0:
            return n
        n += 1

if __name__ == "__main__":
    print(euler_78())  # 55374

File: Java/Problem_078.java
import java.util.*;

public class Euler78 {
    public static void main(String[] args) {
        final int MOD = 1_000_000;

        ArrayList<Integer> p = new ArrayList<>(60000);
        p.add(1); // p(0)

        for (int n = 1; ; n++) {
            long total = 0;

            for (int k = 1; ; k++) {
                int g1 = k * (3 * k - 1) / 2;
                if (g1 > n) break;
                int g2 = k * (3 * k + 1) / 2;

                int sign = (k & 1) == 1 ? 1 : -1;

                total += sign * p.get(n - g1);
                if (g2 <= n) total += sign * p.get(n - g2);
            }

            int val = (int)(total % MOD);
            if (val < 0) val += MOD;

            p.add(val);
            if (val == 0) {
                System.out.println(n); // 55374
                return;
            }
        }
    }
}
